### ISSUE TITLE
Determin which cart to get

### ISSUE THREAD

**Issue by seb86** [co-cart/co-cart#46]
To further improve detection of the cart and its contents based on the user status without requiring to authenticate we need to do an object check.

This check will determine to which version of the cart we have access to. If the cart is not empty or set then we will get the cart contents from the cart class, otherwise, we will get the cart contents from the session.

Help with testing this change will be required.

**Comment by seb86** [comment]
An example of checking the cart contents: `$cart_contents = isset( WC()->cart ) ? WC()->cart->cart_contents : WC()->session->cart;`

**Comment by yaundanns** [comment]
hello. you mean get the cart from session, but i didnt transmit the cookie to the server,  how they know which cart need to be return to client.

**Comment by seb86** [comment]
@yaundanns What cookie are you referring to?

**Comment by yaundanns** [comment]
hello, thanks for your response. the script is working fine. cheers!

At 2018-10-18 10:42:45, "Sébastien Dumont" <notifications@github.com> wrote:


@yaundanns What cookie are you referring to?

**Comment by yaundanns** [comment]
hello, thanks for your response. the script is working fine now. cheers!

**Comment by benhuson** [comment]
Hi, I am testing this API to see if I can get it working for my use case - 1 WordPress site loading products via REST from another WordPress site running WooCommerce and your plugin, and adding the to a cart using your Co Cart API. I have this setup locally as two separate sites.

If I open the site in two separate browsers, both seem to load the same cart and add the same cart although they should be 2 different sessions?

As @yaundanns mentions above, when access via the API how does WooCommerce know which cart need to be return to client as no session/cartid seems to be send with the request?

Is that the reason for this issue?

Thanks

Ben

**Comment by joshuaiz** [comment]
Hi Seb, just bought you a coffee - thanks for the work on this!

I'm running into this issue now. Can you explain your answer a bit more (e.g. 'use anonymous'?). 

We need a way to match the specific cart with the session so that the proper cart is returned. How can I do this with a JavaScript/React front end?

As it is now, I'm getting cart 'ghosting' where I'll have items in the cart from other users' sessions if they don't complete the purchase. I *could* clear the cart on every session locally but that seems a bit extreme.

Also, in my case, users/visitors are not logged in and I don't want to require logins.

If multiple people are accessing the site at once, then the front end app could get confused very quickly.

To me, this is pretty much urgent as I'm seeing the issue pop up a lot now that my site is live: https://vizualrecords.com

**Comment by joshuaiz** [comment]
This is the same issue as #50 and is not the WP API. There needs to be a unique identifier for each cart that is tied to a user session yet this needs to work even if there isn't a server/PHP Session.

Thus, the unique identifier/key for each cart probably should be generated by CoCart and sent along with the product data. This can then be stored in local/session storage or cookie and then sent with every additional request like adding/updating/getting the cart so that the correct cart is updated.

On the client side, we can check if the correct cart is being returned by comparing the key. 

Nevertheless, once the key is created, it should be sent with the endpoint so that we are only grabbing a single cart rather than grabbing *all* carts and then filtering them on the client side which could be slow.

**Comment by seb86** [comment]
> I'm running into this issue now. Can you explain your answer a bit more (e.g. 'use anonymous'?).

Meaning no authentication. Add items to the cart as a guest.

> We need a way to match the specific cart with the session so that the proper cart is returned. How can I do this with a JavaScript/React front end?

I am still learning React myself at the moment so I can't really answer that at this time.

> As it is now, I'm getting cart 'ghosting' where I'll have items in the cart from other users' sessions if they don't complete the purchase. I could clear the cart on every session locally but that seems a bit extreme.

> Also, in my case, users/visitors are not logged in and I don't want to require logins.

Because of the way WooCommerce is designed, support for the cart is lacking the way the API needs it to function which is what I am working on supporting.

> If multiple people are accessing the site at once, then the front end app could get confused very quickly.

I have yet to come across this issue myself locally but I am working on finding a solution.

> Thus, the unique identifier/key for each cart probably should be generated by CoCart and sent along with the product data. This can then be stored in local/session storage or cookie and then sent with every additional request like adding/updating/getting the cart so that the correct cart is updated.

That is what I have come up with and I am in the process of building.

**Comment by joshuaiz** [comment]
Thanks @seb86! 

> Meaning no authentication. Add items to the cart as a guest.

Another idea is if there is no authentication, then WC/WP could create a guest user just for the purposes of the cart so something like `guest1`, `guest2`, etc., and then use that to identify the cart. The users could be destroyed once the purchase is complete. Don't know if that would be easier that creating a key.

We would still have the order data with name/email or whatever data you are capturing on your checkout form.

Anyway, excited to see what you come up with. Thanks!!

**Comment by seb86** [comment]
WooCommerce cart works fine in core because it uses sessions and PHP. However, via an API it needs more support because it is not just data that is being posted or retrieved via the WC API. That is why WC does not support the cart via API.

This is why the first version of CoCart is really just a proof of concept. Hopefully, when I am done with CoCart version 2 I will have added the support the API needs to work with any language. I will put out a beta release for developers to test before hand.

**Comment by joshuaiz** [comment]
Cool that makes sense. 

In the meantime I'll try to work on a workaround using sessionStorage. I think if we can save the users cart there until it's time to purchase and then send that just before the order that could work. A bit hacky but it could work.

**Comment by joshuaiz** [comment]
Just following up on this. Any progress on your end @seb86?

Did a bit of research over the last few days using [JWT (JSON Web Tokens)](https://wordpress.org/plugins/jwt-authentication-for-wp-rest-api/) for authorization as well as looking into everything regarding web sessions + javascript.

What I've found in Postman is that if you make a request to the WooCommerce API using an auth token or using the auth keys provided by WooCommerce, WC returns 3 cookies: `woocommerce_items_in_cart`, `woocommerce_cart_hash`, and `woocommerce_session_xxxxxxxxxxxxx`.

The cookies returned are for the authorized user who is making the request. In Postman, the requests are essentially coming from the same server but in a React/js app, the requests can be coming from anywhere so there is no way to access these cookies or the session data. Of course, this is the main problem we are facing.

Continuing on the client side, I was able to create sort of 'dummy' users easily using the [wp-rest-user](https://github.com/SK8-PTY-LTD/wp-rest-user) plugin so my thought was if we could create a user for each cart, then possibly WC would return the cart from that user. Unfortunately, we face the same issue as there is no way to tie the user to a particular session/cart without some server interaction.

It *might* be possible from the client side if we force users to log in before buying anything but that is not really an option I'd like to pursue.

These were just initial experiments. Now I will try my idea above that will save the cart to a user's `sessionStorage` and then only send it to WooCommerce at the checkout stage. The cart could then be destroyed either after a certain time or if they leave the checkout page. This is still pretty hacky and doesn't solve the issue if multiple people are buying something at the same time.

With all that said, I'm interested to hear any ideas on how the cart + session flow would work and how this could be implemented in a client side js app.

Hopefully @seb86 you've made some progress!
